libbeat/metric/system/process/process_linux_common.go

```diff
@@ -0,0 +1,465 @@
+}
+
+// FetchPids is the linux implementation of FetchPids
+func (procStats *Stats) FetchPids() (ProcsMap, []ProcState, error) {
+	dir, err := os.Open(procStats.Hostfs.ResolveHostFS("proc"))
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "error reading from procfs %s", procStats.Hostfs.ResolveHostFS("/"))
+	}
+	defer dir.Close()
+
+	const readAllDirnames = -1 // see os.File.Readdirnames doc
+
+	names, err := dir.Readdirnames(readAllDirnames)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "error reading directory names")
+	}
+
+	procMap := make(ProcsMap, 0)
+	var plist []ProcState
+
+	// Iterate over the directory, fetch just enough info so we can filter based on user input.
+	logger := logp.L()
+	for _, name := range names {
+
+		if !dirIsPid(name) {
+			continue
+		}
+		// Will this actually fail?
+		pid, err := strconv.Atoi(name)
+		if err != nil {
+			logger.Debugf("Error converting PID name %s", name)
+			continue
+		}
+		procMap, plist = procStats.pidIter(pid, procMap, plist)
+	}
+
+	return procMap, plist, nil
+}
+
+func FillPidMetrics(hostfs resolve.Resolver, pid int, state ProcState, filter func(string) bool) (ProcState, error) {
+	// Memory Data
+	var err error
+	state.Memory, err = getMemData(hostfs, pid)
+	if err != nil {
+		return state, errors.Wrapf(err, "error getting memory data for pid %d", pid)
+	}
+
+	// CPU Data
+	state.CPU, err = getCPUTime(hostfs, pid)
+	if err != nil {
+		return state, errors.Wrapf(err, "error getting CPU data for pid %d", pid)
+	}
+
+	// CLI args
+	if len(state.Args) == 0 {
+		state.Args, err = getArgs(hostfs, pid)
+		if err != nil {
+			return state, errors.Wrapf(err, "error getting CLI args for pid %d", pid)
+		}
+
+	}
+
+	// FD metrics
+	state.FD, err = getFDStats(hostfs, pid)
+	if err != nil {
+		return state, errors.Wrapf(err, "error getting FD metrics for pid %d", pid)
+	}
+
+	if state.Env == nil {
+		// env vars
+		state.Env, err = getEnvData(hostfs, pid, filter)
+	}
+
+	state.Exe, state.Cwd, err = getProcStringData(hostfs, pid)
+	if err != nil {
+		return state, errors.Wrapf(err, "error getting metadata for pid %d", pid)
+	}
+
+	//username
+	state.Username, err = getUser(hostfs, pid)
+	if err != nil {
+		return state, errors.Wrapf(err, "error creating username for pid %d", pid)
+	}
+	return state, nil
+}
+
+// GetInfoForPid fetches the basic hostinfo from /proc/[PID]/stat
+func GetInfoForPid(hostfs resolve.Resolver, pid int) (ProcState, error) {
+	path := hostfs.Join("proc", strconv.Itoa(pid), "stat")
+	data, err := ioutil.ReadFile(path)
+	// Transform the error into a more sensible error in cases where the directory doesn't exist, i.e the process is gone
+	if err != nil {
+		if os.IsNotExist(err) {
+			return ProcState{}, syscall.ESRCH
+		} else {
+			return ProcState{}, errors.Wrapf(err, "error reading procdir %s", path)
+		}
+	}
+
+	state := ProcState{}
+
+	// Extract the comm value with is surrounded by parentheses.
+	lIdx := bytes.Index(data, []byte("("))
```

Comment:
How about using `TrimLeft` and `TrimRight`? To me, it seems a bit easier to read than indexing.

Reply:
Maybe I'm misunderstanding, but we're not trying to remove those particular characters, but indexing to fetch the string inside those characters. The alternative would be regex, and I figure this is a little lower overhead.